File: src/integrationMongoTest/java/io/lsdconsulting/lsd/distributed/interceptor/integration/matcher/InterceptedInteractionMatcher.java
package io.lsdconsulting.lsd.distributed.interceptor.integration.matcher;

import io.lsdconsulting.lsd.distributed.access.model.InteractionType;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import lombok.RequiredArgsConstructor;
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

@RequiredArgsConstructor
public class InterceptedInteractionMatcher extends TypeSafeMatcher<InterceptedInteraction> {

    private final InteractionType type;
    private final String serviceName;
    private final String body;
    private final String target;
    private final String path;

    @Override
    protected boolean matchesSafely(final InterceptedInteraction interceptedInteraction) {
        return interceptedInteraction.getInteractionType().equals(type)
                && interceptedInteraction.getServiceName().equals(serviceName)
                && interceptedInteraction.getBody().equals(body)
                && interceptedInteraction.getTarget().equals(target)
                && interceptedInteraction.getPath().equals(path);
    }

    @Override
    public void describeTo(final Description description) {
        description.appendText("an item with body:" + body + " and serviceName:" + serviceName + " and type:" + type + " and target:" + target + " and path:" + path);
    }

    public static InterceptedInteractionMatcher with(final InteractionType type, final String serviceName, final String body, final String target, final String path) {
        return new InterceptedInteractionMatcher(type, serviceName, body, target, path);
    }
}

File: src/integrationMongoTest/kotlin/io/lsdconsulting/lsd/distributed/interceptor/integration/matcher/InterceptedInteractionMatcher.kt
package io.lsdconsulting.lsd.distributed.interceptor.integration.matcher

import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import org.hamcrest.Description
import org.hamcrest.TypeSafeMatcher

class InterceptedInteractionMatcher(
    private val type: InteractionType,
    private val serviceName: String,
    private val body: String,
    private val target: String,
    private val path: String,
) : TypeSafeMatcher<InterceptedInteraction>() {

    override fun matchesSafely(interceptedInteraction: InterceptedInteraction): Boolean {
        return interceptedInteraction.interactionType == type && interceptedInteraction.serviceName == serviceName && interceptedInteraction.body == body && interceptedInteraction.target == target && interceptedInteraction.path == path
    }

    override fun describeTo(description: Description) {
        description.appendText("an item with body:$body and serviceName:$serviceName and type:$type and target:$target and path:$path")
    }

    companion object {
        fun with(
            type: InteractionType,
            serviceName: String,
            body: String,
            target: String,
            path: String
        ): InterceptedInteractionMatcher {
            return InterceptedInteractionMatcher(type, serviceName, body, target, path)
        }
    }
}